### PR TITLE
[Feature] New RBAC framework: support basic db & table actions

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Introduce a new privilege framework.
To avoid huge PR which is hard to review and easy to conflict, I'm splitting into several small PR, each implementing an individual aspect of the new privilege framework. This is the sixth one. For previous PRs see 
https://github.com/StarRocks/starrocks/pull/11447
https://github.com/StarRocks/starrocks/pull/11528
https://github.com/StarRocks/starrocks/pull/11593
https://github.com/StarRocks/starrocks/pull/11732
https://github.com/StarRocks/starrocks/pull/11777

This PR implements basic authorization of db and table:
* create_table/drop for database.
* select/insert/delete for tables.
A minor modification: extract the logic of analyzing action/type/object from executer to analyzer.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 96 / 103 (93.20%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/privilege/PEntryObject.java|0|1|00.00%|[23]|
|:large_blue_circle:|com/starrocks/privilege/TablePEntryObject.java|2|3|66.67%|[39]|
|:large_blue_circle:|com/starrocks/privilege/DbPEntryObject.java|11|13|84.62%|[29, 35]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|19|21|90.48%|[201, 275]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|23|24|95.83%|[40]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeTypes.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/DefaultAuthorizationProvider.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|10|10|100.00%|[]|

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 131 / 140 (93.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/privilege/PEntryObject.java|1|3|33.33%|[24, 41]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|1|2|50.00%|[767]|
|:large_blue_circle:|com/starrocks/privilege/DbPEntryObject.java|12|14|85.71%|[29, 35]|
|:large_blue_circle:|com/starrocks/privilege/TablePEntryObject.java|6|7|85.71%|[39]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeCollection.java|14|15|93.33%|[190]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|23|24|95.83%|[40]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|34|35|97.14%|[203]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeTypes.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/DefaultAuthorizationProvider.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 131 / 140 (93.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/privilege/PEntryObject.java|1|3|33.33%|[24, 41]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|1|2|50.00%|[767]|
|:large_blue_circle:|com/starrocks/privilege/DbPEntryObject.java|12|14|85.71%|[29, 35]|
|:large_blue_circle:|com/starrocks/privilege/TablePEntryObject.java|6|7|85.71%|[39]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeCollection.java|14|15|93.33%|[190]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|23|24|95.83%|[40]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|34|35|97.14%|[218]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeTypes.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/DefaultAuthorizationProvider.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|10|10|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 1]
```diff
@@ -760,7 +760,11 @@ public void addTabletToInvertedIndex() {
                 }
             } // end for indices
         }
+    }
 
+    public void removeInvalidateReference() {
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:765]: Where will this function be called

Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:765]: Good point! I forgot...just added the caller few seconds ago.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 133 / 141 (94.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/privilege/PEntryObject.java|1|3|33.33%|[24, 41]|
|:large_blue_circle:|com/starrocks/privilege/DbPEntryObject.java|12|14|85.71%|[29, 35]|
|:large_blue_circle:|com/starrocks/privilege/TablePEntryObject.java|6|7|85.71%|[39]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeCollection.java|14|15|93.33%|[190]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|23|24|95.83%|[40]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|34|35|97.14%|[218]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeTypes.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/DefaultAuthorizationProvider.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|10|10|100.00%|[]|

**Review by gengjun-git — DISMISSED** [review]

**Review by waittttting — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 2]
```diff
@@ -3367,8 +3367,9 @@ public ParseNode visitRevokeImpersonateBrief(StarRocksParser.RevokeImpersonateBr
 
     @Override
     public ParseNode visitGrantTablePrivBrief(StarRocksParser.GrantTablePrivBriefContext context) {
+        int objectTokenSize = context.tablePrivilegeObjectName().identifierOrStringOrStar().size();
         return visitGrantRevokePrivWithType(
-                "TABLE",
+                objectTokenSize == 1 ? "DATABASE" : "TABLE",
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:3372]: Is this too tricky? Shall we add a new function like `visitGrantDatabasePrivBrief()`?

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 3]
```diff
@@ -760,7 +760,13 @@ public void addTabletToInvertedIndex() {
...
+        if (GlobalStateMgr.getCurrentState().isUsingNewPrivilege()) {
+            // privilege object can be invalidated after gc
+            GlobalStateMgr.getCurrentState().getPrivilegeManager().removeInvalidateObject();
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java]: done

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java [hunk 4]
```diff
@@ -289,13 +299,35 @@ private ActionSet checkActionSet(String typeName, short typeId, List<String> act
...
+    public void removeInvalidateObject() {
+        Iterator<Map.Entry<UserIdentity, UserPrivilegeCollection>> mapIter =
+                userToPrivilegeCollection.entrySet().iterator();
+        while (mapIter.hasNext()) {
+            Map.Entry<UserIdentity, UserPrivilegeCollection> entry = mapIter.next();
+            UserIdentity user = entry.getKey();
+            UserPrivilegeCollection collection = entry.getValue();
+            if (! globalStateMgr.getAuthenticationManager().doesUserExist(user)) {
+                String collectionStr = GsonUtils.GSON.toJson(collection);
+                LOG.info("find invalidate user {}, will remove privilegeCollection now {}",
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java]: done

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java [hunk 5]
```diff
@@ -642,7 +642,7 @@ public Void visitGrantRevokePrivilegeStatement(BaseGrantRevokePrivilegeStmt stmt
                     ErrorReport.reportSemanticException(ErrorCode.ERR_SPECIFIC_ACCESS_DENIED_ERROR, "GRANT");
                 }
             } else {
-                if (stmt.getPrivType().equals("TABLE")) {
+                if (stmt.getPrivType().equals("TABLE") || stmt.getPrivType().equals("DATABASE")) {
```
[the remark this answers is not in the document]
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java:645]: It's a legacy reason. The first one is a "table level"  statement and the second one is a "database level" statement. In the previous implementation, these two statements are handled in the same logical path, but in the current implementation, it is not.
```sql
grant select on dbx.tablex to userx
grant select on dbx.* to userx
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java (hunk 2 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:3372]: I think there are too many functions in this class...I prefer not to add more unless really necessary.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java:645]: got it

**Review by nshangyiming — DISMISSED** [review]

**Review by gengjun-git — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 136 / 144 (94.44%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/privilege/PEntryObject.java|1|3|33.33%|[24, 41]|
|:large_blue_circle:|com/starrocks/privilege/DbPEntryObject.java|12|14|85.71%|[29, 35]|
|:large_blue_circle:|com/starrocks/privilege/TablePEntryObject.java|6|7|85.71%|[39]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeCollection.java|14|15|93.33%|[190]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|23|24|95.83%|[40]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|34|35|97.14%|[218]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeTypes.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/DefaultAuthorizationProvider.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|10|10|100.00%|[]|

**Review by gengjun-git — APPROVED** [review]

**Review by nshangyiming — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Pull request merged**